Implement iter_all_for_block and iter_all_for_course for DjangoUserStateClient

clemente opened the pull request:
This implements:
- `DjangoXBlockUserStateClient.iter_all_for_block`
- `DjangoXBlockUserStateClient.iter_all_for_course`

They return an iterator over the data inside an XBlock and a course, respectively.
Tests from https://github.com/edx/edx-user-state-client/blob/master/edx_user_state_client/tests.py#L603 still work with this implementation. 

**JIRA tickets**: None
**Project details**: This PR is part of the "Improved Problem Response Report" project, funded by Australian National University (ANU) and developed by OpenCraft.                                                                                                                   
**Discussions**: None
**Dependencies**: It depends on version 1.0.4 of edx-user-state-client, which must be released after merging this PR: https://github.com/edx/edx-user-state-client/pull/14
**Screenshots**: None
**Sandbox URL**: None
**Merge deadline**: None

**Testing instructions**:

1. Run tests

**Reviewers**
- [ ] @bradenmacdonald 
- [ ] edX's

**Author concerns**: Pagination was set to a high number to avoid high load on the data source (which may potentially be remote). But this doesn't affect the results because results are returned 1 by 1.

**Settings**: None

edx-webhook commented:
Thanks for the pull request, @clemente! I've created [OSPR-2301](https://openedx.atlassian.net/browse/OSPR-2301) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

nedbat commented:
@edx/platform-team Take a look.

clemente commented:
@macdiesel: makes sense, `batch_size` seems some internal performance setting. My plan is to keep the parameter and make the 5000 a setting in `settings`. But another option would be remove the parameter from the implementation and [the interface](https://github.com/edx/edx-user-state-client/blob/master/edx_user_state_client/interface.py#L245), though this would need a version bump at edx-user-state.
I'll schedule this for next week.

bradenmacdonald commented:
We should probably remove it from the interface then too, to reduce confusion. I don't think it's necessary.

clemente commented:
@macdiesel I moved the `batch_size` parameter into settings.
I also removed it from the interface: could you approve+merge https://github.com/edx/edx-user-state-client/pull/14 first, then tag 1.0.4, then recheck the new commit in this PR?
I'll rebase and squash commits then.

clemente commented:
The tests didn't want to start running, so I did a rebase now.

clemente commented:
Tests will fail until edx-user-state-client 1.0.4 is released.

macdiesel commented:
jenkins run all

macdiesel commented:
@clemente edx-user-state-client 1.0.4 has been released. I have restarted the tests. ping me once you have squashed and rebased and I'll merge.

clemente commented:
@macdiesel Thanks, I squashed and rebased this started the tests, I'll ping you when they pass.
The last tests still didn't find 1.0.4

macdiesel commented:
@clemente you may want to rebase and try again, this is a conflict in the requirements file now as anyway.

clemente commented:
@macdiesel I rebased and now tests pass.

edx-webhook commented:
@clemente 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/17698&entry.1484655318=https://github.com/clemente&entry.752974735=2018-03-15+12:04&entry.1917517419=2018-04-20+12:44&entry.2133058324=Yes) so we can improve your experience in the future.

clemente commented:
@macdiesel  Thanks!
